cdap-ui/app/cdap/components/DataPrep/Directives/Calculate/index.js

```diff
@@ -250,30 +271,36 @@ export default class Calculate extends Component {
   }
 
   getExpressionAndApply() {
-    let expression;
+    if (this.isApplyDisabled()) {
+      return;
+    }
 
-    switch (this.state.operationPopoverOpen) {
-      case 'ADD':
-        expression = `${this.props.column} + ${this.state.operationInput}`;
-        break;
-      case 'SUBTRACT':
-        expression = `${this.props.column} - ${this.state.operationInput}`;
-        break;
-      case 'MULTIPLY':
-        expression = `${this.props.column} * ${this.state.operationInput}`;
-        break;
-      case 'DIVIDE':
-        expression = `${this.props.column} / ${this.state.operationInput}`;
-        break;
-      case 'MODULO':
-        expression = `${this.props.column} % ${this.state.operationInput}`;
-        break;
-      case 'POWEROF':
-        expression = `math:pow(${this.props.column}, ${this.state.operationInput})`;
-        break;
-      case 'CHARCOUNT':
-        expression = `string:length(${this.props.column})`;
-        break;
+    let expression;
+    let operationObj = this.CALCULATE_OPTIONS.filter(option => this.state.operationPopoverOpen === option.name)[0];
+
+    if (operationObj && operationObj.expression) {
+      expression = operationObj.expression;
+    } else {
+      switch (this.state.operationPopoverOpen) {
+        case 'ADD':
+          expression = `${this.props.column} + ${this.state.operationInput}`;
```

Comment:
If we have `expression` as function that returns the expression we could avoid the `switch` entirely. Just execute the expression function all the time to get the expression. That will help evaluating the string expression only when it is needed (which will pick up the correct value of `state.operationInput`.